data/transactions/logic/eval.go

```diff
@@ -905,16 +902,15 @@ func (cx *EvalContext) step() error {
 		return fmt.Errorf("%3d %s program ends short of immediate values", cx.pc, spec.Name)
 	}
 
-	opcost := deets.Cost
-	if opcost == 0 {
-		// We use a constant (deets.Cost) to avoid the cost of an indirect
-		// function call in most cases.  Consider getting the cost details into
-		// OpSpec so that a single (inlinable?) function can compute the cost,
-		// rather than a per opcode function pointer.
-		opcost = deets.costFunc(cx.program, cx.pc)
-	}
-	if opcost == 0 {
-		return fmt.Errorf("%3d %s returned 0 cost", cx.pc, spec.Name)
+	// It's something like a 5-10% overhead on our simplest instructions to make
```

Comment:
It's my impression that until 1.18, Go would *never* inline a method with a for loop (maybe just range for loops?), so that explains what's going on.  We could look again in 1.18 as those recent blog posts imply that some inlining of such methods are doable.

Reply:
@jannotti Thanks for the replies - addresses my questions.  I leave responses for completeness below and I'm resolving the thread.

> It's my impression that until 1.18, Go would never inline a method with a for loop 

* Correct - https://go.dev/doc/go1.18#compiler shows v1.18 enriches inlining capabilities.
* In my local changes, I split `Cost` implementation as shown below hoping to reduce its cost enough to support inlining the duplicated check.  Based on my read of https://dave.cheney.net/2020/05/02/mid-stack-inlining-in-go, provided the inlining budget isn't exceeded, it's a valid inlining attempt.

```
func (d opDetails) Cost(program []byte, pc int, stack []stackValue) int {
	cost := d.FullCost.compute(stack)

	if cost != 0 {
		return cost
	}

	return d.ImmediateCost(program, pc)
}

func (d opDetails) ImmediateCost(program []byte, pc int) int {
	cost := 0
	for i := range d.Immediates {
		if d.Immediates[i].fieldCosts != nil {
			cost += d.Immediates[i].fieldCosts[program[pc+1+i]]
		}
	}
	return cost
}
```

>  I've never looked into at all.

Got it - I set a reminder to investigate the continuous benchmarking out-of-band to the PR.